test/common.js

```diff
@@ -354,3 +354,61 @@ exports.fileExists = function(pathname) {
     return false;
   }
 };
+
+
+
+(function() {
+  var asyncTestsEnabled = false;
+  var asyncTestLastCheck;
+  var asyncTestQueue = [];
+  var asyncTestHandle;
+  var currentTest = null;
+
+  function fail(error) {
+    var stack = currentTest
+          ? error.stack + "\nFrom previous event:\n" + currentTest.stack
+          : error.stack;
+
+    process.stderr.write(stack);
+    process.exit(2);
+  }
+
+  function nextAsyncTest() {
+    var called = false;
+    function done(err) {
+      if (called) return fail(new Error("done called twice"));
+      called = true;
+      asyncTestLastCheck = Date.now();
+      if (arguments.length > 0) return fail(err);
+      setTimeout(nextAsyncTest, 10);
+    }
+
+    if (asyncTestQueue.length) {
+      var test = asyncTestQueue.shift();
+      currentTest = test;
+      test.action(done);
+    } else {
+      clearInterval(asyncTestHandle);
+      process.exit(0);
+    }
+  }
+
+  exports.asyncTest = function(fn) {
+    var stack = new Error().stack.split("\n").slice(1).join("\n");
+    asyncTestQueue.push({action: fn, stack: stack});
+    if (!asyncTestsEnabled) {
+      asyncTestsEnabled = true;
+      asyncTestLastCheck = Date.now();
+      process.on("uncaughtException", fail);
+      asyncTestHandle = setInterval(function() {
+        var now = Date.now();
+        if (now - asyncTestLastCheck > 10000) {
+          return fail(new Error("Async test timeout exceeded"));
+        }
+      }, 10);
+      setTimeout(nextAsyncTest, 10);
```

Comment:
Using setInterval() and setTimeout() with a hard-coded timeout seems incredibly fragile to me.  Ditto for the fixed 10 second total running time.